plugin.program.AML/resources/utils_kodi.py

```diff
@@ -0,0 +1,220 @@
+# Kodi notifications and dialogs
+# -----------------------------------------------------------------------------
+#
+# Displays a modal dialog with an OK button. Dialog can have up to 3 rows of text, however first
+# row is multiline.
+# Call examples:
+#  1) ret = kodi_dialog_OK('Launch ROM?')
+#  2) ret = kodi_dialog_OK('Launch ROM?', title = 'AML - Launcher')
+#
+def kodi_dialog_OK(row1, row2='', row3='', title = 'Advanced MAME Launcher'):
+    dialog = xbmcgui.Dialog()
+    dialog.ok(title, row1, row2, row3)
+
+#
+# Returns True is YES was pressed, returns False if NO was pressed or dialog canceled.
+def kodi_dialog_yesno(row1, row2='', row3='', title = 'Advanced MAME Launcher'):
+    dialog = xbmcgui.Dialog()
+    ret = dialog.yesno(title, row1, row2, row3)
+
+    return ret
+
+#
+# Displays a small box in the bottom right corner
+#
+def kodi_notify(text, title = 'Advanced MAME Launcher', time = 5000):
+    dialog = xbmcgui.Dialog()
+    dialog.notification(title, text, xbmcgui.NOTIFICATION_INFO, time)
+
+def kodi_notify_warn(text, title = 'Advanced MAME Launcher warning', time = 7000):
+    dialog = xbmcgui.Dialog()
+    dialog.notification(title, text, xbmcgui.NOTIFICATION_WARNING, time)
+
+#
+# Do not use this function much because it is the same icon as when Python fails, and that may confuse the user.
+#
+def kodi_notify_error(text, title = 'Advanced MAME Launcher error', time = 7000):
+    dialog = xbmcgui.Dialog()
+    dialog.notification(title, text, xbmcgui.NOTIFICATION_ERROR, time)
+
+def kodi_busydialog_ON():
+    xbmc.executebuiltin('ActivateWindow(busydialog)')
+
+def kodi_busydialog_OFF():
+    xbmc.executebuiltin('Dialog.Close(busydialog)')
+
+def kodi_refresh_container():
+    log_debug('kodi_refresh_container()')
+    xbmc.executebuiltin('Container.Refresh')
+
+# -----------------------------------------------------------------------------
+# Kodi specific stuff
+# -----------------------------------------------------------------------------
+# About Kodi image cache
+#
+# See http://kodi.wiki/view/Caches_explained
+# See http://kodi.wiki/view/Artwork
+# See http://kodi.wiki/view/HOW-TO:Reduce_disk_space_usage
+# See http://forum.kodi.tv/showthread.php?tid=139568 (What are .tbn files for?)
+#
+# Whenever Kodi downloads images from the internet, or even loads local images saved along 
+# side your media, it caches these images inside of ~/.kodi/userdata/Thumbnails/. By default, 
+# large images are scaled down to the default values shown below, but they can be sized 
+# even smaller to save additional space.
+
+#
+# Gets where in Kodi image cache an image is located. 
+#
+def kodi_get_cached_image(image_path):
+    THUMBS_CACHE_PATH = os.path.join(xbmc.translatePath('special://profile/' ), 'Thumbnails')
+
+    # --- Get the Kodi cached image ---
+    # This function return the cache file base name
+    base_name = xbmc.getCacheThumbName(image_path)
```

Comment:
Why you need to access thumbnails directly?

Reply:
That is legacy code from Advanced Launcher. The purpose was to force an update of Kodi image cache. When you scrape some piece of artwork the image changes in the filesystem, however Kodi keeps showing the old cached image for a time and the effect is rather annoying. AML does not use that code because is does not have online scrapers, currently that code is used by Advanced Emulator Launcher only. Note that `utils.py` and `utils_kodi.py` are shared between both addons. I will remove that code.

Do you know what is the proper way to tell Kodi to force the re-cache of a cached image that has changed? If there is no a proper way I will open a request in the forum.

Reply:
If it is not used please remove or at least comment the code block. If you need kodi to store a new thumbnail you should use point the listitem to a new image (e.g. append a timestamp to the image file name)

Reply:
The code has been removed. AML does not have online scrapers so it wasn't used anyway. I will investigate the image file timestamps in AEL. Thanks a lot for the tip!